pkgs/development/libraries/catboost/default.nix

```diff
@@ -0,0 +1,111 @@
+{ lib
+, config
+, stdenv
+, fetchFromGitHub
+, cmake
+, libiconv
+, llvmPackages
+, ninja
+, openssl
+, python3
+, ragel
+, yasm
+, zlib
+, cudaSupport ? config.cudaSupport or false
+, cudatoolkit
+, enablePython ? false
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "catboost";
+  version = "1.2";
+
+  src = fetchFromGitHub {
+    owner = "catboost";
+    repo = "catboost";
+    rev = "refs/tags/v${finalAttrs.version}";
+    hash = "sha256-Ng1ylqL9fBRJP5hATv9FzJTqqSIFEUN0At1DJ3V2FM4=";
+  };
+
+  patches = [
+    ./remove-conan.patch
+    ./fix-imported-targets.patch
+  ] ++ lib.optionals (lib.versionOlder cudatoolkit.version "11.8") [
+    ./remove-unsupport-cc.patch
+  ];
+
+  postPatch = ''
+    substituteInPlace cmake/common.cmake \
+      --replace  "\''${RAGEL_BIN}" "${lib.getExe ragel}" \
+      --replace "\''${YASM_BIN}" "${lib.getExe yasm}"
+  '';
+
+  outputs = [
+    # binary and libraries e.g. bin/catboost, lib/catboost.so, etc.
+    "out"
+    # libraries required for Python, R, etc. support
+    # see https://catboost.ai/en/docs/installation/build-native-artifacts#targets
+    "dev"
+  ];
+
+  nativeBuildInputs = [
+    cmake
+    llvmPackages.bintools
+    ninja
+    ragel
+    yasm
+  ];
+
+  buildInputs = [
+    openssl
+    zlib
+  ] ++ lib.optionals stdenv.isDarwin [
+    libiconv
+  ] ++ lib.optionals cudaSupport [
+    (cudatoolkit.override { backendStdenv = stdenv; })
+  ];
+
+  env = {
+    CUDAHOSTCXX = "${stdenv.cc}/bin/clang";
```

Comment:
@NixOS/cuda-maintainers 
In case `backendStdenv = clangStdenv`, it needs to set `CUDAHOSTCXX = ${stdenv.cc}/bin/clang` explicitly.
This is because `-DCUDA_HOST_COMPILER=${backendStdenv.cc}/bin` is set here.
https://github.com/NixOS/nixpkgs/blob/3ae20aa58a6c0d1ca95c9b11f59a2d12eebc511f/pkgs/development/compilers/cudatoolkit/common.nix#L272-L276
Since nvcc looks for gcc by default, it tries to find `clang/bin/gcc` without the CUDAHOSTCXX.
Is it worth setting `-DCUDA_HOST_COMPILER=${lib.getExe backendStdenv.cc.cc}` in cudatoolkit?

Reply:
Hi! Could you give some more intro on why this needs to be built with clang?

The reason the `backendStdenv` attribute is there is because different cudatoolkit versions support different backend compiler releases (e.g. nvcc from cudatoolkit 11.7 will simply error out when used with gcc12). At the moment we only maintain a table with nvcc-gcc versions compatibility, but in principle one could add llvm as well. The `(cudatoolkit.override { backendStdenv = clangStdenv; })` might work but only accidentally and until nixpkgs bumps the default llvm major version.

> `-DCUDA_HOST_COMPILER=${lib.getExe backendStdenv.cc.cc}`

There are two issues with this line:
1. Both the legacy `CUDA_HOST_COMPILER` and the `CMAKE_CUDA_HOST_COMPILER` map into nvcc's `-ccbin`, which expects a directory. Iirc it'd use `${ccbin}/cc` or `${ccbin}/cxx` as the actual compiler. I don't remember exactly, but I think `CUDAHOSTCXX` works the same
2. The un-wrapped `backendStdenv.cc.cc` (opposed to `backendStdenv.cc`) is something you probably don't want to use. E.g. `cudaPackages_11.backendStdenv.cc.cc` is gcc11 and it links its artifacts to gcc11's libstdc++, which won't work with the rest of nixpkgs. Similarly, if you were using clang, you'd need to wrap it so that it would link gcc's `libstdc++` instead of llvm's `libcxx`. I'm sorry about this mess, there's an open issue tracking this here: https://github.com/NixOS/nixpkgs/issues/226165

Reply:
Also, you probably want to use `cudaPackages.cuda_{nvcc,cudart,..}` instead of `cudatoolkit`

Reply:
@SomeoneSerge san, Thanks for the clarification.
I need clang12+ to build this catboost and can't build it with GCC.
ref: https://catboost.ai/en/docs/installation/build-environment-setup-for-cmake#compilers,-linkers-and-related-tools
I understand the role of backendStdenv, and even if accidentally, I should fix the clang version here.

> Both the legacy CUDA_HOST_COMPILER and the CMAKE_CUDA_HOST_COMPILER map into nvcc's -ccbin, which expects a directory. Iirc it'd use ${ccbin}/cc or ${ccbin}/cxx as the actual compiler. I don't remember exactly, but I think CUDAHOSTCXX works the same

Yes, `-ccbin` expects a directory, but the [documentation](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#compiler-bindir-directory-ccbin) says it is possible to specify an executable.
The `backendStdenv=clangStdenv` case is a problem because it tries to run `${ccbin}/gcc` instead of `${ccbin}/cc` as mentioned above.
In fact, building this PR change without CUDAHOSTCXX fails.

> The un-wrapped backendStdenv.cc.cc (opposed to backendStdenv.cc) is something you probably don't want to use. E.g. cudaPackages_11.backendStdenv.cc.cc is gcc11 and it links its artifacts to gcc11's libstdc++, which won't work with the rest of nixpkgs. Similarly, if you were using clang, you'd need to wrap it so that it would link gcc's libstdc++ instead of llvm's libcxx. I'm sorry about this mess, there's an open issue tracking this here: cudaPackages: extend cc-wrapper and rewrite cudaStdenv to properly solve libstdc++ issues 

Thanks for letting me know about the cc-wrapper issue.
I will change it to link libstdc++.

Reply:
Btw, if you'd like to prevent cudatoolkit/cuda_nvcc from exporting CUDAHOSTCXX&c, you should be able to use `dontSetupCUDAToolkitCompilers = true` once https://github.com/NixOS/nixpkgs/pull/233581/files#diff-412326f1eb2e17a306d14d70da7fe2423b4f244aa3c9c81d13f1bb1cf3be084bR20 is merged

Reply:
Hello again! You mentioned earlier that

> [Since] nvcc looks for gcc by default [...].

I may be now facing a similar issue with cross-compilation, so I was wondering if you really meant nvcc (rather than cmake, or even findcuda/findcudatoolkit specifically), and if you've got a reference documenting this behaviour?

Reply:
Sorry, I can't answer this question for sure, but I guess it is the behavior of nvcc.
Looking at CMake's FindCUDAToolkit and other CUDA-related sources, it does not seem to refer specifically to gcc.
I can't find any documentation on this behavior, but JetBrain's CLion page says the following for reference.
https://www.jetbrains.com/help/clion/cuda-projects.html#custom-host-compiler
> For compiling host code, NVCC calls the system's default C++ compiler (gcc/g++ on Linux and cl.exe on Windows).